samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -0,0 +1,145 @@
+ *   {@link org.I0Itec.zkclient.IZkStateListener} and subscribing this listener to the current ZkClient. Note: The connection state change
+ *   callbacks are invoked in the context of the Event thread of the ZkClient. So, it is advised to do non-blocking
+ *   processing in the callbacks.
+ * </p>
+ */
+public class ZkUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(ZkUtils.class);
+
+  private final ZkClient zkClient;
+  private volatile String ephemeralPath = null;
+  private final ZkKeyBuilder keyBuilder;
+  private final int connectionTimeoutMs;
+
+  public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs) {
+    this.keyBuilder = zkKeyBuilder;
+    this.connectionTimeoutMs = connectionTimeoutMs;
+    this.zkClient = zkClient;
+  }
+
+  public void connect() throws ZkInterruptedException {
+    boolean isConnected = zkClient.waitUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS);
+    if (!isConnected) {
+      throw new RuntimeException("Unable to connect to Zookeeper within connectionTimeout " + connectionTimeoutMs + "ms. Shutting down!");
+    }
+  }
+
+  public static ZkConnection createZkConnection(String zkConnectString, int sessionTimeoutMs) {
+    return new ZkConnection(zkConnectString, sessionTimeoutMs);
+  }
+
+  public static ZkClient createZkClient(ZkConnection zkConnection, int connectionTimeoutMs) {
+    return new ZkClient(zkConnection, connectionTimeoutMs);
+  }
+
+  ZkClient getZkClient() {
+    return zkClient;
+  }
+
+  public ZkKeyBuilder getKeyBuilder() {
+    return keyBuilder;
+  }
+
+  /**
+   * Returns a ZK generated identifier for this client.
+   * If the current client is registering for the first time, it creates an ephemeral sequential node in the ZK tree
+   * If the current client has already registered and is still within the same session, it returns the already existing
+   * value for the ephemeralPath
+   *
+   * @param data Object that should be written as data in the registered ephemeral ZK node
+   * @return String representing the absolute ephemeralPath of this client in the current session
+   */
+  public synchronized String registerProcessorAndGetId(final Object data) {
+    if (ephemeralPath == null) {
+      // TODO: Data should be more than just the hostname. Use Json serialized data
+      ephemeralPath =
+          zkClient.createEphemeralSequential(keyBuilder.getProcessorsPath() + "/processor-", data);
+      return ephemeralPath;
+    } else {
+      return ephemeralPath;
+    }
+  }
+
+  public synchronized String getEphemeralPath() {
+    return ephemeralPath;
+  }
+
+  /**
+   * Method is used to get the <i>sorted</i> list of currently active/registered processors
+   *
+   * @return List of absolute ZK node paths
+   */
+  public List<String> getActiveProcessors() {
+    List<String> children = zkClient.getChildren(keyBuilder.getProcessorsPath());
+    if (children.size() > 0) {
+      Collections.sort(children);
```

Comment:
The idea behind ZkUtils is to make any Zookeeper access transparent to the caller. Otherwise, the caller needs to know that the ZkClient/Zookeeper API (for getChildren) does not return an ordered list of children.  I don't see any benefit of returning an unordered list of processors. Is there any reason why you think the caller should sort as opposed to the data accessing class?